addons/libs/dialog.lua

```diff
@@ -0,0 +1,165 @@
+-- This library was written to help find the ID of a known
+-- action message corresponding to an entry in the dialog tables.
+-- While the IDs can be collected in-game, they occasionally
+-- change and would otherwise need to be manually updated.
+-- It can also be used to find and decode an entry given the ID.
+
+-- Common parameters:
+-- dat: The entire content of the zone dialog DAT file
+-- i.e. local dat = io.open('path/to/dialog/DAT', 'rb'):read('*a')
+-- entry: The string you are looking for. If you do not know the
+-- entire string, use dev_find_substring.
+
+local xor = require('bit').bxor
+local floor = require('math').floor
+local string = require('string')
+local find = string.find
+local sub = string.sub
+local gsub = string.gsub
+local format = string.format
+local char = string.char
+local byte = string.byte
+require('pack')
+local unpack = string.unpack
+local pack = string.pack
+
+local dialog = {}
+
+local function decode(int)
+    return xor(int, 0x80808080)
+end
+local encode = decode
+
+local function binary_search(pos, dat, n)
+    local l, r, m = 1, n
+    while l < r do
+        m = floor((l + r) / 2)
+        if decode(unpack('<I', dat, 1 + 4 * m)) < pos then
+        -- offset given by mth ID < offset to string
+            l = m + 1
+        else
+            r = m
+        end
+    end
+    return l - 2 -- we want the index to the left of where "pos" would be placed
+end
+
+local function plain_text_gmatch(text, substring, n)
+    n = n or 1
+    return function()
+        local pos = find(text, substring, n, true)
+        if pos then n = pos + 1 end
+        return pos
+    end
+end
+
+-- This function may help if you are viewing the decoded contents
+-- of a dialog DAT within a hex editor.
+function dialog.dev_get_offset(dat, id)
+    return format('%x', decode(unpack('<I', dat, 5 + 4 * id)))
```

Comment:
I'd prefer a prefixed and fixed-size offset display like `'0x%8X'`, but not sure if your intention was to use it for some specific tool that expects it in a different format.

Reply:
I was using it to pipe the offset to my clipboard so that I could paste it into a hex editor. It wouldn't be difficult to trim the first two characters, though, so that's cool with me. I added it as `'0x%08X'`